File: python/concurrency/ex2_shared_state.py
"""
Exercise 2: Shared State and Synchronization (Medium)

Learn how to safely share state between threads using synchronization.
This exercise covers:
- Lock and RLock for thread safety
- Condition variables for coordination
- Thread-safe data structures

Difficulty: Medium
Topics: Thread Safety, Locks, Synchronization, Race Conditions
"""

import threading
from typing import Optional


class SynchronizedCounter:
    """
    Task 1: Thread-safe counter using Lock

    TODO: Use threading.Lock to make operations thread-safe
    """

    def __init__(self):
        self.count = 0
        # TODO: Create a lock
        self.lock = None

    def increment(self):
        # TODO: Acquire lock, increment, release lock
        # Hint: Use 'with self.lock:' context manager
        raise NotImplementedError("Not implemented yet")

    def decrement(self):
        # TODO: Make this thread-safe
        raise NotImplementedError("Not implemented yet")

    def get_count(self) -> int:
        # TODO: Make this thread-safe
        raise NotImplementedError("Not implemented yet")


class BankAccount:
    """
    Task 2: Thread-safe bank account

    TODO: Implement thread-safe deposit, withdraw, and transfer operations
    """

    def __init__(self, initial_balance: float):
        self.balance = initial_balance
        self.lock = threading.Lock()

    def deposit(self, amount: float):
        # TODO: Thread-safe deposit
        raise NotImplementedError("Not implemented yet")

    def withdraw(self, amount: float) -> bool:
        # TODO: Thread-safe withdraw, return False if insufficient funds
        raise NotImplementedError("Not implemented yet")

    def get_balance(self) -> float:
        # TODO: Thread-safe balance retrieval
        raise NotImplementedError("Not implemented yet")

    def transfer_to(self, other: 'BankAccount', amount: float) -> bool:
        # TODO: Thread-safe transfer between accounts
        # Hint: Need to lock both accounts - avoid deadlock by ordering locks
        # Hint: Use id(self) and id(other) to determine lock order
        raise NotImplementedError("Not implemented yet")


class SharedBuffer:
    """
    Task 3: Shared buffer with wait/notify pattern

    TODO: Implement put/take with Condition variable
    """

    def __init__(self):
        self.value = None
        self.has_value = False
        self.condition = threading.Condition()

    def put(self, item):
        # TODO: Wait while buffer has value, then put new value
        # Hint: Use self.condition.wait() and self.condition.notify_all()
        raise NotImplementedError("Not implemented yet")

    def take(self):
        # TODO: Wait while buffer is empty, then take value
        raise NotImplementedError("Not implemented yet")


class ThreadSafeSingleton:
    """
    Task 4: Thread-safe singleton pattern

    TODO: Implement double-checked locking
    """

    _instance = None
    _lock = threading.Lock()

    def __new__(cls):
        # TODO: Implement double-checked locking pattern
        # Hint: Check instance without lock, then with lock
        raise NotImplementedError("Not implemented yet")


class ReadWriteLock:
    """
    Task 5: Read-write lock implementation

    Multiple readers can access simultaneously, but writers get exclusive access.

    TODO: Implement using Condition variables
    """

    def __init__(self):
        self.readers = 0
        self.writers = 0
        self.condition = threading.Condition()

    def acquire_read(self):
        # TODO: Acquire read lock (wait if writer is active)
        raise NotImplementedError("Not implemented yet")

    def release_read(self):
        # TODO: Release read lock
        raise NotImplementedError("Not implemented yet")

    def acquire_write(self):
        # TODO: Acquire write lock (wait if any readers or writers)
        raise NotImplementedError("Not implemented yet")

    def release_write(self):
        # TODO: Release write lock
        raise NotImplementedError("Not implemented yet")


if __name__ == "__main__":
    print("Concurrency - Exercise 2: Shared State\n")

    # Example: Demonstrating race condition
    print("Without synchronization (race condition):")

    class UnsafeCounter:
        def __init__(self):
            self.count = 0

        def increment(self):
            self.count += 1

    unsafe = UnsafeCounter()
    threads = []
    for _ in range(10):
        t = threading.Thread(target=lambda: [unsafe.increment() for _ in range(1000)])
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    print(f"Expected: 10000, Actual: {unsafe.count}")
    print("(Likely less than 10000 due to race condition)\n")
    print("Implement the TODO tasks to make operations thread-safe!")

File: rust/concurrency/src/ex2_shared_state.rs
/// Exercise 2: Shared State (Medium)
///
/// Learn to share mutable state safely between threads.
/// This exercise covers:
/// - Mutex<T> for mutual exclusion
/// - Arc<T> for shared ownership
/// - RwLock<T> for read-write locking
///
/// Difficulty: Medium
/// Topics: Mutex, Arc, RwLock, Interior Mutability

use std::sync::{Arc, Mutex, RwLock};
use std::thread;

/// Task 1: Thread-safe counter using Mutex
///
/// TODO: Implement thread-safe counter with Arc<Mutex<i32>>
pub fn task1_mutex_counter(num_threads: usize, increments_per_thread: usize) -> i32 {
    // TODO: Create Arc<Mutex<i32>>, clone for each thread, increment
    unimplemented!("Not implemented yet")
}

/// Task 2: Shared vector with Mutex
///
/// TODO: Multiple threads add items to shared Vec
pub fn task2_shared_vector(num_threads: usize) -> Vec<i32> {
    // TODO: Share Vec<i32> with Arc<Mutex<...>>
    unimplemented!("Not implemented yet")
}

/// Task 3: Read-write lock for cache
///
/// TODO: Use RwLock for read-heavy workload
/// Hint: RwLock allows multiple readers OR one writer
pub fn task3_rwlock_cache() -> String {
    // TODO: Create Arc<RwLock<String>>, spawn readers and writers
    unimplemented!("Not implemented yet")
}

/// Task 4: Bank account transfers
///
/// TODO: Implement thread-safe transfers between accounts
pub struct BankAccount {
    balance: Arc<Mutex<f64>>,
}

impl BankAccount {
    pub fn new(initial: f64) -> Self {
        // TODO: Initialize with Arc<Mutex<...>>
        unimplemented!("Not implemented yet")
    }

    pub fn deposit(&self, amount: f64) {
        // TODO: Lock and add to balance
        unimplemented!("Not implemented yet")
    }

    pub fn withdraw(&self, amount: f64) -> bool {
        // TODO: Lock, check balance, withdraw if sufficient
        unimplemented!("Not implemented yet")
    }

    pub fn balance(&self) -> f64 {
        // TODO: Return current balance
        unimplemented!("Not implemented yet")
    }
}

/// Task 5: Deadlock-free transfer
///
/// TODO: Transfer between accounts without deadlock
/// Hint: Use consistent lock ordering
pub fn transfer(from: &BankAccount, to: &BankAccount, amount: f64) -> bool {
    // TODO: Implement deadlock-free transfer
    unimplemented!("Not implemented yet")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_mutex_counter() {
        let count = task1_mutex_counter(10, 1000);
        assert_eq!(count, 10000);
    }

    #[test]
    fn test_shared_vector() {
        let vec = task2_shared_vector(10);
        assert_eq!(vec.len(), 10);
    }

    #[test]
    fn test_rwlock_cache() {
        let result = task3_rwlock_cache();
        assert!(!result.is_empty());
    }

    #[test]
    fn test_bank_account() {
        let account = BankAccount::new(1000.0);
        account.deposit(500.0);
        assert_eq!(account.balance(), 1500.0);
        assert!(account.withdraw(200.0));
        assert_eq!(account.balance(), 1300.0);
    }

    #[test]
    fn test_transfer() {
        let acc1 = BankAccount::new(1000.0);
        let acc2 = BankAccount::new(500.0);
        assert!(transfer(&acc1, &acc2, 300.0));
        assert_eq!(acc1.balance(), 700.0);
        assert_eq!(acc2.balance(), 800.0);
    }
}
